tools/pyang_plugins/protobuf.py

```diff
@@ -0,0 +1,647 @@
+
+            out.write('\n')
+
+        if self.rpcs:
+            self._print_rpc(out)
+
+        return out
+
+
+class YangContainer(object):
+    def __init__(self):
+        self.name = None
+        self.containers = []
+        self.enums = []
+        self.leafs = []
+        self.ylist = []
+
+class YangList(object):
+    def __init__(self):
+        self.name = None
+        self.leafs = []
+        self.containers = []
+        self.ylist = []
+
+
+class YangLeaf(object):
+    def __init__(self):
+        self.name = None
+        self.type = None
+        self.json_name = None
+        self.leaf_list = False
+        self.enumeration = []
+        self.enumeration_names = set()
+        self.description = None
+
+
+class YangEnumeration(object):
+    def __init__(self):
+        self.value = []
+
+
+class YangRpc(object):
+    def __init__(self):
+        self.mod_name = None
+        self.name = None
+        self.name_without_parent = None
+        self.input = 'google.protobuf.Empty'
+        self.output = 'google.protobuf.Empty'
+        self.input_empty = False
+        self.output_empty = False
+        self.url = None
+
+
+class ProtobufPlugin(plugin.PyangPlugin):
+    def add_output_format(self, fmts):
+        self.multiple_modules = True
+        fmts['proto'] = self
+
+    def setup_fmt(self, ctx):
+        ctx.implicit_errors = False
+
+    def add_opts(self, optparser):
+        optlist = [
+            optparse.make_option("--proto-outdir",
+                                 type="string",
+                                 dest="outdir",
+                                 help="Output directory for protobuffs"),
+            optparse.make_option("--server-rpc-outdir",
+                                 type="string",
+                                 dest="server_stub_outdir",
+                                 help="Output directory for server stubs"),
+            optparse.make_option("--client-rpc-outdir",
+                                 type="string",
+                                 dest="client_outdir",
+                                 help="Output directory for server stubs"),
+        ]
+        g = optparser.add_option_group("OpenApiPlugin options")
+        g.add_options(optlist)
+
+    def write_file(self, fn, content):
+        fileChanged = True
+        if os.path.isfile(fn):
+            with open(fn) as fp:
+                fileChanged = (fp.read() != content)
+        if fileChanged:
+            with open(fn, "w") as fp:
+                print(f"writing file: {fn}")
+                fp.write(content)
+        else:
+            print(f"file {fn} unchanged, skipped writing...")
+        return fileChanged
+
+    def emit(self, ctx, modules, fd):
+        """Main control function.
+        """
+        global current_proto
+        self.ctx = ctx
+        for idx, d in enumerate([ctx.opts.outdir, ctx.opts.server_stub_outdir, ctx.opts.client_outdir]):
+            if not d:
+                if idx == 0:
```

Comment:
Lets make more concise and readable by using f-strings.
```
            print(f"--proto-outdir cannot be empty" if idx == 0 else
                  f"--server-rpc-outdir cannot be empty" if idx == 1 else
                  "--client-rpc-outdir cannot be empty")
```

Reply:
I think current one is more readable. but its subjective, do you see any computational benefit?